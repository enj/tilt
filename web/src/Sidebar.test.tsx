import React from "react"
import { mount } from "enzyme"
import renderer from "react-test-renderer"
import { MemoryRouter, useHistory } from "react-router"
import Sidebar, { SidebarItemLink } from "./SidebarResources"
import SidebarItem from "./SidebarItem"
import { oneResource, twoResourceView } from "./testdata"
import { ResourceView, TriggerMode } from "./types"
import PathBuilder from "./PathBuilder"
import SidebarResources from "./SidebarResources"
import fetchMock from "jest-fetch-mock"

type Resource = Proto.webviewResource

let pathBuilder = new PathBuilder("localhost", "/")

let realDateNow = Date.now

describe("sidebar", () => {
  beforeEach(() => {
    fetchMock.resetMocks()
    Date.now = jest.fn(() => Date.UTC(2017, 11, 21, 15, 36, 7, 0))
  })
  afterEach(() => {
    Date.now = realDateNow
  })
  it("renders empty resource list without crashing", () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <SidebarResources
            items={[]}
            selected=""
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders list of resources", () => {
    let items = twoResourceView().resources.map((res: Resource) => {
      let history = res.buildHistory ?? []
      history[0].error = "error!"
      return new SidebarItem(res)
    })
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <SidebarResources
            items={items}
            selected=""
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("abbreviates durations under a minute", () => {
    let items = [4, 9, 19, 29, 39, 49, 54].map(d => {
      let res = oneResource()
      res.name = `resource${d}`
      res.lastDeployTime = new Date(Date.now() - d * 1000).toISOString()
      return new SidebarItem(res)
    })

    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <SidebarResources
            items={items}
            selected=""
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("renders resources that haven't been built yet", () => {
    let items = twoResourceView().resources.map((res: any) => {
      // currently building, no completed builds
      res.lastDeployTime = "0001-01-01T00:00:00Z"
      return new SidebarItem(res)
    })
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/"]}>
          <SidebarResources
            items={items}
            selected=""
            resourceView={ResourceView.Log}
            pathBuilder={pathBuilder}
          />
        </MemoryRouter>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("reloads selected resource on click", () => {
    let items = twoResourceView().resources.map(
      (res: any) => new SidebarItem(res)
    )
    const root = mount(
      <MemoryRouter initialEntries={["/"]}>
        <SidebarResources
          items={items}
          selected={items[0].name}
          resourceView={ResourceView.Log}
          pathBuilder={pathBuilder}
        />
      </MemoryRouter>
    )

    let links = root.find(SidebarItemLink)
    expect(links).toHaveLength(3)

    links.at(1).simulate("click")

    expect(fetchMock.mock.calls.length).toEqual(1)
    expect(fetchMock.mock.calls[0][0]).toEqual("/api/trigger")
    expect(fetchMock.mock.calls[0][1]?.method).toEqual("post")
    expect(fetchMock.mock.calls[0][1]?.body).toEqual(
      JSON.stringify({
        manifest_names: ["vigoda"],
        build_reason: 16 /* BuildReasonFlagTriggerWeb */,
      })
    )

    root.unmount()
  })

  it("navigates to unselected resource on click", () => {
    let items = twoResourceView().resources.map(
      (res: any) => new SidebarItem(res)
    )
    let history: any
    let CaptureHistory = () => {
      history = useHistory()
      return <span />
    }
    const root = mount(
      <MemoryRouter initialEntries={["/"]}>
        <CaptureHistory />
        <SidebarResources
          items={items}
          selected={items[0].name}
          resourceView={ResourceView.Log}
          pathBuilder={pathBuilder}
        />
      </MemoryRouter>
    )

    let links = root.find(SidebarItemLink)
    expect(links).toHaveLength(3)
    expect(links.at(2).html()).toContain('href="/r/snack"')
    root.unmount()
  })
})
