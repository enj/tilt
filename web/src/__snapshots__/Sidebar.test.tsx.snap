// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar abbreviates durations under a minute 1`] = `
<nav
  className="sc-hKFyIo eUFJqs Sidebar-resources"
>
  <ul
    className="sc-eCApGN ha-dpKs"
  >
    <li
      className="sc-jSFkmK sc-gKAblj gNevMY jRtnvI isSelected"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/"
        onClick={[Function]}
        title="all"
      >
        <div
          className="sc-bdnylx bqmyzl isNone"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            7
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          All
        </p>
      </a>
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource4"
        onClick={[Function]}
        title="resource4"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource4
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:36:03.000Z"
              title="2017-12-21T15:36:03.000Z"
            >
              &lt;5s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource9"
        onClick={[Function]}
        title="resource9"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource9
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:58.000Z"
              title="2017-12-21T15:35:58.000Z"
            >
              &lt;15s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource19"
        onClick={[Function]}
        title="resource19"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource19
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:48.000Z"
              title="2017-12-21T15:35:48.000Z"
            >
              &lt;30s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource29"
        onClick={[Function]}
        title="resource29"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource29
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:38.000Z"
              title="2017-12-21T15:35:38.000Z"
            >
              &lt;30s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource39"
        onClick={[Function]}
        title="resource39"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource39
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:28.000Z"
              title="2017-12-21T15:35:28.000Z"
            >
              &lt;45s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource49"
        onClick={[Function]}
        title="resource49"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource49
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:18.000Z"
              title="2017-12-21T15:35:18.000Z"
            >
              &lt;1m ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/resource54"
        onClick={[Function]}
        title="resource54"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            resource54
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:13.000Z"
              title="2017-12-21T15:35:13.000Z"
            >
              &lt;1m ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
  </ul>
  <span />
</nav>
`;

exports[`sidebar renders empty resource list without crashing 1`] = `
<nav
  className="sc-hKFyIo eUFJqs Sidebar-resources"
>
  <ul
    className="sc-eCApGN ha-dpKs"
  >
    <li
      className="sc-jSFkmK sc-gKAblj gNevMY jRtnvI isSelected"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/"
        onClick={[Function]}
        title="all"
      >
        <div
          className="sc-bdnylx bqmyzl isNone"
        >
          <span>
             
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          All
        </p>
      </a>
    </li>
  </ul>
  <span />
</nav>
`;

exports[`sidebar renders list of resources 1`] = `
<nav
  className="sc-hKFyIo eUFJqs Sidebar-resources"
>
  <ul
    className="sc-eCApGN ha-dpKs"
  >
    <li
      className="sc-jSFkmK sc-gKAblj gNevMY jRtnvI isSelected"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/"
        onClick={[Function]}
        title="all"
      >
        <div
          className="sc-bdnylx bqmyzl isNone"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            2
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          All
        </p>
      </a>
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/vigoda"
        onClick={[Function]}
        title="vigoda"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            vigoda
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:36:07.000Z"
              title="2017-12-21T15:36:07.000Z"
            >
              &lt;5s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            12s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/snack"
        onClick={[Function]}
        title="snack"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            snack
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG"
          >
            <time
              dateTime="2017-12-21T15:35:57.000Z"
              title="2017-12-21T15:35:57.000Z"
            >
              &lt;15s ago
            </time>
          </span>
          <span
            className="sc-jrsJCI fhJJUT"
          >
            -10.0s
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
  </ul>
  <span />
</nav>
`;

exports[`sidebar renders resources that haven't been built yet 1`] = `
<nav
  className="sc-hKFyIo eUFJqs Sidebar-resources"
>
  <ul
    className="sc-eCApGN ha-dpKs"
  >
    <li
      className="sc-jSFkmK sc-gKAblj gNevMY jRtnvI isSelected"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/"
        onClick={[Function]}
        title="all"
      >
        <div
          className="sc-bdnylx bqmyzl isNone"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            2
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          All
        </p>
      </a>
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/vigoda"
        onClick={[Function]}
        title="vigoda"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            vigoda
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG isEmpty"
          >
            —
          </span>
          <span
            className="sc-jrsJCI fhJJUT isEmpty"
          >
            —
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
    <li
      className="sc-jSFkmK gNevMY  isBuilding"
    >
      <a
        className="sc-crzoUp gMPNfP"
        href="/r/snack"
        onClick={[Function]}
        title="snack"
      >
        <div
          className="sc-bdnylx bqmyzl isBuilding"
        >
          <span
            className="sc-gtssRu bdlHqP"
          >
            1
          </span>
        </div>
        <p
          className="sc-iCoHVE dwxClQ"
        >
          <span
            className="sc-fujyUd fIrNil"
          >
            snack
          </span>
        </p>
        <div
          className="sc-pNWxx yWHfW"
        >
          <span
            className="sc-kEqYlL eJIBeG isEmpty"
          >
            —
          </span>
          <span
            className="sc-jrsJCI fhJJUT isEmpty"
          >
            —
          </span>
        </div>
      </a>
      <button
        className="sc-dlnjPT bGTNQz SidebarTriggerButton 
          "
        disabled={true}
        onClick={[Function]}
        title="Cannot trigger an update while resource is updating or update is pending."
      />
    </li>
  </ul>
  <span />
</nav>
`;
